### PR TITLE
ReactNotificationService to allow communication between native modules

### PR DESCRIPTION
**Pull request by vmoroz**
This PR introduces a new ReactNotificationService that allows native modules to communicate between each other, communicate with view managers, and with the application.
One examples of the usage is when a native module wants to listen to UIManager events. Instead of receiving UIManager as a parameter, the native module can just subscribe to the UIManager events in its `Initialize` method using the `ReactContext::Notifications()` property.

The `IReactNotificationService` has two methods:
- `Subscribe` to subscribe to notification. It receives `IReactPropertyName` as a name of the notification (we reuse the PropertyBag's atomized names in this context), an optional `IReactDispatcher` where the notification must be handled, and the handler to handle notification.  The `Subscribe` method returns `IReactNotificationSubscription` that has the subscription name, dispatcher, and `Unsubscribe` method to unsubscribe from the notification. Its `IsSubscribed` property can be used to check if subscription is still active.
- `SendNotification` to send a notification. It gets the notification name, sender object, and notification data object. The sender and data objects are optional and can be null.

The notification handler receives sender object (`IInspectable` in C++) and `IReactNotificationArgs` that has `Subscription` and `Data` properties. The `Subscription` property has the same value as the result of the `Subscribe` method call and can be used to get the subscription properties and to be able to unsubscribe inside of the handler. The `Data` property provides data sent by the `SendNotificaton` data.

On top of the ABI-safe API, we offer wrappers for all new types: `ReactNotificationService`, `ReactNotificationSubscription`, `ReactNotificationArgs`, `ReactNotificationId` and `ReactDispatcher`.
The `ReactNotificationId` is a template that allows to specify an ABI-safe type to be used as notification data. The type can be `void`. For non-ABI safe value in local context we can use `ReactNonAbiValue<T>` type.

The code below is taken from the sample apps to show usage of the new ReactNotificationService:
```C++
  REACT_INIT(Initialize)
  void Initialize(ReactContext const &reactContext) noexcept {
    const ReactNotificationId<int> cppTimerNotification{L"SampleModuleCppImpl", L"TimerNotification"};
    const ReactNotificationId<int> csTimerNotification{L"SampleModuleCS", L"TimerNotification"};

    reactContext.Notifications().Subscribe(csTimerNotification,
       [ ](winrt::Windows::Foundation::IInspectable const &, ReactNotificationArgs<int> const &args) noexcept {
      DEBUG_OUTPUT("C++ module, C# timer:", *args.Data());
    });

    m_timer = winrt::Windows::System::Threading::ThreadPoolTimer::CreatePeriodicTimer(
        [ this, cppTimerNotification, notifications = reactContext.Notifications() ](
            const winrt::Windows::System::Threading::ThreadPoolTimer) noexcept {
          notifications.SendNotification(cppTimerNotification, m_timerCount);
          TimedEvent(++m_timerCount);
          if (m_timer && m_timerCount == 5) {
            m_timer.Cancel();
          }
        },
        TimedEventInterval);
  }
```
```C#
        [ReactInitializer]
        public void Initialize(ReactContext reactContext)
        {
            var cppTimerNotification = ReactPropertyBagHelper.GetName(ReactPropertyBagHelper.GetNamespace("SampleModuleCppImpl"), "TimerNotification");
            var csTimerNotification = ReactPropertyBagHelper.GetName(ReactPropertyBagHelper.GetNamespace("SampleModuleCS"), "TimerNotification");

            reactContext.Handle.Notifications.Subscribe(cppTimerNotification, null,
                (object sender, IReactNotificationArgs args) => Debug.WriteLine($"C# module, C++ timer:: {args.Data}"));

            _timer = ThreadPoolTimer.CreatePeriodicTimer(new TimerElapsedHandler((timer) =>
            {
                reactContext.Handle.Notifications.SendNotification(csTimerNotification, null, _timerCount);
                TimedEvent?.Invoke(++_timerCount);
                if (_timerCount == 5)
                {
                    _timer?.Cancel();
                }
            }),
            TimeSpan.FromMilliseconds(TimedEventIntervalMS));
        }
```
The C# usage looks verbose. We will add helper classes in future PR.

The next examples are taken from our unit tests.
This is how we can unsubscribe from the notifications:
```C++
    ReactNotificationService rns{ReactNotificationServiceHelper::CreateNotificationService()};
    ReactNotificationId<void> fooNotification{L"Foo"};
    bool isCalled{false};
    auto subscription = rns.Subscribe(
        fooNotification, [&](IInspectable const & /*sender*/, ReactNotificationArgs<void> const & /*args*/) noexcept {
          isCalled = true;
        });
    rns.SendNotification(fooNotification);
    TestCheck(isCalled);

    subscription.Unsubscribe();
    TestCheck(!subscription.IsSubscribed());

    isCalled = false;
    rns.SendNotification(fooNotification);
    TestCheck(!isCalled);
```

We can use the `winrt::auto_revoke` value to get `ReactNotificationSubscriptionRevoker` instead of `ReactNotificationSubscription`:
```C++
    ReactNotificationService rns{ReactNotificationServiceHelper::CreateNotificationService()};
    ReactNotificationId<void> fooNotification{L"Foo"};
    bool isCalled{false};
    auto revoker = rns.Subscribe(winrt::auto_revoke, fooNotification,
        [&](IInspectable const & /*sender*/, ReactNotificationArgs<void> const & /*args*/) noexcept {
          isCalled = true;
        });
    rns.SendNotification(fooNotification);
    TestCheck(isCalled);

    revoker = nullptr; // destruction of revoker causes unsubscription.
    TestCheck(!subscription.IsSubscribed());

    isCalled = false;
    rns.SendNotification(fooNotification);
    TestCheck(!isCalled);
```

We can unsubscribe from inside of the the handler:
```C++
    ReactNotificationService rns{ReactNotificationServiceHelper::CreateNotificationService()};
    ReactNotificationId<void> fooNotification{L"Foo"};
    bool isCalled{false};
    auto subscription = rns.Subscribe(
        fooNotification, [&](IInspectable const & /*sender*/, ReactNotificationArgs<void> const &args) noexcept {
          isCalled = true;
          args.Subscription().Unsubscribe();
        });
    rns.SendNotification(fooNotification);
    TestCheck(isCalled);

    isCalled = false;
    rns.SendNotification(fooNotification);
    TestCheck(!isCalled);
```

We can provide a `ReactDispatcher` where to handle notification:
```C++
    ReactNotificationService rns{ReactNotificationServiceHelper::CreateNotificationService()};
    ReactNotificationId<ReactNonAbiValue<std::string>> fooNotification{L"Foo"};
    Mso::ManualResetEvent finishedEvent;
    ReactDispatcher dispatcher{ReactDispatcher::CreateSerialDispatcher()};
    bool isCalled{false};
    rns.Subscribe(fooNotification, dispatcher,
        [&](IInspectable const & /*sender*/,
            ReactNotificationArgs<ReactNonAbiValue<std::string>> const &args) noexcept {
          TestCheckEqual("Hello", *args.Data());
          TestCheckEqual(dispatcher, args.Subscription().Dispatcher());
          TestCheck(dispatcher.HasThreadAccess());
          isCalled = true;
          finishedEvent.Set();
        });
    rns.SendNotification(fooNotification, "Hello");
    finishedEvent.Wait();
    TestCheck(isCalled);
```

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/4953)

### CODE REVIEW

**Comment by NikoAri** [comment]
> namespace SampleLibraryCpp {

I'm probably missing something really obvious, but is this header public? Why wouldn't this functionality be in its own file? 

---
Refers to: packages/microsoft-reactnative-sampleapps/windows/SampleLibraryCPP/SampleModuleCPP.h:15 in 696d8c6. [](commit_id = 696d8c65716cc1950b0a201e6f1a3c34125506a7, deletion_comment = False)

File: vnext/Microsoft.ReactNative/IReactNotificationService.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  IReactNotificationSubscription m_subscription;
+  IInspectable m_data;
+};
+
+struct ReactNotificationService : implements<ReactNotificationService, IReactNotificationService> {
+  ReactNotificationService();
+  explicit ReactNotificationService(IReactNotificationService const parentNotificationService) noexcept;
+  ~ReactNotificationService() noexcept;
+
+  IReactNotificationSubscription Subscribe(
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/IReactNotificationService.h:38]: >Subscribe [](start = 33, length = 9)

Just curious, we have CallbackStore for events in devmain, would it make sense to use something like that here?

File: vnext/Microsoft.ReactNative/IReactNotificationService.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  ReactNotificationSubscription(
+      weak_ref<ReactNotificationService> &&notificationService,
+      IReactPropertyName const &notificationName,
+      IReactDispatcher const &dispatcher,
+      ReactNotificationHandler const &handler) noexcept;
+  ~ReactNotificationSubscription() noexcept;
+
+  IReactPropertyName NotificationName() const noexcept;
+  IReactDispatcher Dispatcher() const noexcept;
+  bool IsSubscribed() const noexcept;
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/IReactNotificationService.h:83]: >IsSubscribed [](start = 7, length = 12)

Why is this needed? Wouldn't it create race condition if someone calls if (IsSubscribed()) call()? (but someone unsubscribes in-between)

**Comment by NikoAri** [comment]
>     IReactPropertyNamespace Namespace { get; };

(nit) why isn't this just a string?

---
Refers to: vnext/Microsoft.ReactNative/IReactPropertyBag.idl:28 in 696d8c6. [](commit_id = 696d8c65716cc1950b0a201e6f1a3c34125506a7, deletion_comment = False)

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h [hunk 3]
```diff
@@ -41,16 +41,21 @@ class ReactContext final : public Mso::UnknownObject<IReactContext> {
  public:
   ReactContext(
       Mso::WeakPtr<ReactInstanceWin> &&reactInstance,
-      winrt::Microsoft::ReactNative::IReactPropertyBag const &properties) noexcept;
+      winrt::Microsoft::ReactNative::IReactPropertyBag const &properties,
+      winrt::Microsoft::ReactNative::IReactNotificationService const &notifications) noexcept;
+
+  void Destroy() noexcept;
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h:47]: >void [](start = 2, length = 4)

Can you add quick comment on why should users need this as opposed to just having destructor taking care of this?

Comment (NikoAri) [inline vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h:47]: >Destroy [](start = 7, length = 7)

Can you add quick comment on why should users need this as opposed to just having destructor taking care of this?

File: vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace winrt::Microsoft::ReactNative {
+
+// Represents a dispatcher queue to invoke work item asynchronously.
+// It wraps up the IReactDispatcher and adds convenience methods for
+// working with C++ types.
+struct ReactDispatcher {
+  ReactDispatcher(std::nullptr_t = nullptr) noexcept {}
+
+  explicit ReactDispatcher(IReactDispatcher const &handle) noexcept : m_handle{handle} {}
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h:19]: > & [](start = 49, length = 2)

(nit) Since we are doing assignment anyways, wouldn't && be better here?

File: vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  IReactDispatcher const &Handle() const noexcept {
+    return m_handle;
+  }
+
+  explicit operator bool() const noexcept {
+    return static_cast<bool>(m_handle);
+  }
+
+  void Post(ReactDispatcherCallback const &callback) const noexcept {
+    if (m_handle) {
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h:30]: >m_handle [](start = 8, length = 8)

(nit) can you add comment of why we would ever want this to be null?

File: vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    return m_handle ? m_handle.IsSubscribed() : false;
+  };
+
+  // Remove the subscription.
+  // Because of the multi-threaded nature of the notifications, the handler can be still called
+  // after the Unsubscribe method called if the IsSubscribed property is already checked.
+  // Consider calling the Unsubscribe method and the handler in the same IReactDispatcher
+  // to ensure that no handler is invoked after the Unsubscribe method call.
+  void Unsubscribe() const noexcept {
+    if (m_handle) {
```
Comment (NikoAri) [inline vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h:58]: >if (m_handle) { [](start = 4, length = 15)

Why would we ever allow null handle? Wouldn't it be easier if m_handle is always an object?

File: vnext/Microsoft.ReactNative/IReactNotificationService.h (hunk 2 above)
Comment (NikoAri) [reply vnext/Microsoft.ReactNative/IReactNotificationService.h:83]: Another way of putting it, what information does return value from this function give? If after returning false, there may still be callbacks en-route.

---
In reply to: [427621090](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427621090) [](ancestors = 427621090)

**Comment by NickGerleman** [comment]
@vmoroz I know we talked offline about this a bit, but I don't think this should be targeting 0.62. It's pretty much out the door, and we won't be able to iterate on this like we've done for other recent APIs.

**Comment by asklar** [comment]
is it by design that the notification service is per react context?
should components from different react contexts be able to communicate?

File: vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  ReactDispatcher(std::nullptr_t = nullptr) noexcept {}
+
+  explicit ReactDispatcher(IReactDispatcher const &handle) noexcept : m_handle{handle} {}
+
+  IReactDispatcher const &Handle() const noexcept {
+    return m_handle;
+  }
+
+  explicit operator bool() const noexcept {
+    return static_cast<bool>(m_handle);
```
Comment (asklar) [inline vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h:26]: consider making this explicit by comparing against nullptr instead

File: vnext/Microsoft.ReactNative.Cxx/ReactHandleHelper.h [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+// True if two types with Handle() have the same handle.
+template <class T, std::enable_if_t<HasHandleV<T>, int> = 0>
+inline bool operator==(T const &left, T const &right) noexcept {
+  return left.Handle() == right.Handle();
+}
+
+// True if two types with Handle() have different handles.
+template <class T, std::enable_if_t<HasHandleV<T>, int> = 0>
+inline bool operator!=(T const &left, T const &right) noexcept {
+  return left.Handle() != right.Handle();
```
Comment (asklar) [inline vnext/Microsoft.ReactNative.Cxx/ReactHandleHelper.h:38]: consider not using the != operator, and instead use  `!(left.Handle() == right.Handle())`

File: vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+#ifndef MICROSOFT_REACTNATIVE_REACTNOTIFICATIONSERVICE
+#define MICROSOFT_REACTNATIVE_REACTNOTIFICATIONSERVICE
+
+#include "ReactDispatcher.h"
+#include "ReactPropertyBag.h"
+
+namespace winrt::Microsoft::ReactNative {
+
+// Encapsulates the IReactPropertyName and the notification data type
+template <class T>
```
Comment (asklar) [inline vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h:14]: consider simplifying this by just making this an alias instead of inheriting from it

File: vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+    return *this;
+  }
+
+  ~ReactNotificationSubscriptionRevoker() noexcept {
+    Unsubscribe();
+  }
+};
+
+struct ReactNotificationArgsBase {
+  ReactNotificationArgsBase(std::nullptr_t = nullptr) noexcept {}
```
Comment (asklar) [inline vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h:92]: mark the ctor as protected so we don't create this from outside the subclasses?

File: vnext/Microsoft.ReactNative.Cxx/ReactPropertyBag.h [hunk 11]
```diff
@@ -98,35 +92,34 @@ struct ReactPropertyNamespace {
...
-// Encapsulates the IReactPropertyName and the property type
-template <class T>
-struct ReactPropertyId {
-  using PropertyType = T;
-
-  ReactPropertyId(std::nullptr_t = nullptr) noexcept {}
+// ReactPropertyName encapsulates the IReactPropertyName.
+// It represents an atomic property name object that defines a LocalName
+// within the referenced Namespace.
+struct ReactPropertyName {
```
Comment (asklar) [inline vnext/Microsoft.ReactNative.Cxx/ReactPropertyBag.h:110]: the name Name would seem to hint that this is the leaf name i.e. the local name (whereas `Id` can be global). Curious to understand the reasoning behind the rename?

**Comment by vmoroz** [comment]
In the current design we have two levels of Notifications: the top level is in the ReactInstanceSettings.Notifications. They have the same lifespan as the ReactInstanceSettings which is usually given to ReactNativeHost. Then we create a new ReactContext.Notifications that use the ReactInstanceSettings.Notifications as parent. The SendNotification will use them both, while subscribe goes only to ReactContext.Notifications. All ReactContext.Notifications are automatically unsubscribed when we unload the React instance and destroy the ReactContext. This design allows modules and views to talk to each other, and the ReactInstanceSettings.Notifications opens communication with the hosting application.

Re: Should components from different react contexts be able to communicate?
- It is not possible in the current implementation. We can add it if we have such scenario.

---
In reply to: [631132307](https://github.com/microsoft/react-native-windows/pull/4953#issuecomment-631132307) [](ancestors = 631132307)

**Comment by vmoroz** [comment]
I believe this API is quite important for the 0.62 release as we plan to implement a lot of new native modules. 

---
In reply to: [631128574](https://github.com/microsoft/react-native-windows/pull/4953#issuecomment-631128574) [](ancestors = 631128574)

File: vnext/Microsoft.ReactNative/IReactNotificationService.h (hunk 1 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/IReactNotificationService.h:38]: This implementation uses a lot of ideas from the CallbackStore. But, it has some more advanced features such as the IReactSubscription object. It is much more powerful when retuning just a 'cookie'

---
In reply to: [427618875](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427618875) [](ancestors = 427618875)

File: vnext/Microsoft.ReactNative/IReactNotificationService.h (hunk 2 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/IReactNotificationService.h:83]: It is useful inside of the notification handler. We check this property before we call the notification handler, but like you said the notification could be in-route. The only thread-safe to ensure that no code executed after unsubscription is to do another check inside of handler. It should be done only for special cases involved multi-threading, but at least we allow it.
In normal cases developers do not need to check this property.

---
In reply to: [427632411](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427632411) [](ancestors = 427632411,427621090)

File: vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h (hunk 4 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h:19]: It is the C++/WinRT thing. It always gives us const &.

---
In reply to: [427624661](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427624661) [](ancestors = 427624661)

File: vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h (hunk 5 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h:30]: This class supports move semantic. It can be null if its content is moved out, or nullptr is given at construction. 

---
In reply to: [427625320](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427625320) [](ancestors = 427625320)

File: vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h (hunk 6 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h:58]: See the other answer and the constructor that accepts nullptr.

---
In reply to: [427626890](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427626890) [](ancestors = 427626890)

File: vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h (hunk 7 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactDispatcher.h:26]: I will change it to code that C++/WinRT uses:
m_handle ? true : false;

Comparing with nullptr is more expensive as it involves calling a constructor with nullptr parameter and then comparing two objects.

---
In reply to: [427648694](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427648694) [](ancestors = 427648694)

File: vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h (hunk 9 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h:14]: I do not see how it can be done. The alias will not allow me to access the template type. Inheritance here is cheap: we inherit from a non-template class.

---
In reply to: [427649718](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427649718) [](ancestors = 427649718)

File: vnext/Microsoft.ReactNative.Cxx/ReactPropertyBag.h (hunk 11 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactPropertyBag.h:110]: I need to reuse it for the ReactNotificationId. Plus, it became a little bit more efficient since we have less method implementations that depend on T.

---
In reply to: [427651462](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427651462) [](ancestors = 427651462)

**Comment by vmoroz** [comment]
>     IReactPropertyNamespace Namespace { get; };

Because IReactPropertyNamespace is an atomized object that groups IReactPropertyName instances.

---
In reply to: [631103652](https://github.com/microsoft/react-native-windows/pull/4953#issuecomment-631103652) [](ancestors = 631103652)

---
Refers to: vnext/Microsoft.ReactNative/IReactPropertyBag.idl:28 in 696d8c6. [](commit_id = 696d8c65716cc1950b0a201e6f1a3c34125506a7, deletion_comment = False)

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h (hunk 3 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.h:50]: Added the comment.

---
In reply to: [427623454](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427623454) [](ancestors = 427623454)

File: vnext/Microsoft.ReactNative.Cxx/ReactHandleHelper.h (hunk 8 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactHandleHelper.h]: Changed

---
In reply to: [427649165](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427649165) [](ancestors = 427649165)

File: vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h (hunk 10 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/ReactNotificationService.h]: They were made protected.

---
In reply to: [427650276](https://github.com/microsoft/react-native-windows/pull/4953#discussion_r427650276) [](ancestors = 427650276)

**Comment by NickGerleman** [comment]
We're not backporting additional payload to 0.62 without triage at this time.

**Comment by vmoroz** [comment]
Stop removing my tags!

---
In reply to: [631223708](https://github.com/microsoft/react-native-windows/pull/4953#issuecomment-631223708) [](ancestors = 631223708)

**Comment by NickGerleman** [comment]
> Stop removing my tags!
> 
> In reply to: [631223708](https://github.com/microsoft/react-native-windows/pull/4953#issuecomment-631223708) [](ancestors = 631223708)

Discussed this more offline. We have been using tags to record what we've been backporting. We talked about this on Monday, and a bit more after, but we're not auto backporting anything more into 0.62. 

As we discussed offline, we can create an issue (not PR) for triage to discuss this. Full guidelines are coming out. Apologies again for all of the confusion on this process 😊.

**Comment by NickGerleman** [comment]
> > Stop removing my tags!
> > In reply to: [631223708](https://github.com/microsoft/react-native-windows/pull/4953#issuecomment-631223708) [](ancestors = 631223708)
> 
> Discussed this more offline. We have been using tags to record what we've been backporting. We talked about this on Monday, and a bit more after, but we're not auto backporting anything more into 0.62.
> 
> As we discussed offline, we can create an issue (not PR) for triage to discuss this. Full guidelines are coming out. Apologies again for all of the confusion on this process 😊.

Talked to @Khalef1 about this offline. You shouldn't need to create an issue for this, and he will drive the triage process for this item.

**Comment by Khalef1** [comment]
Discussed this offline. @vmoroz it would be good to discuss this one at triage. I'll send you the invite.

**Comment by kaiguo** [comment]
Thanks @vmoroz for adding this. About back porting, the [reanimated module](https://github.com/microsoft/react-native-windows/issues/4151) I'm working on will need this to listen on some UIManager events (just providing some more data points for you guys to consider when having the triage discussion 😃).

**Comment by NickGerleman** [comment]
Based on earlier discussion during triage it sounded like we wanted to wait until we had the UIManager part of this, and a better understanding of timing of dependents before considering backporting, Updating labels on this PR. @vmoroz please feel free to re-add the labels if you believe this is incorrect.

**Comment by Khalef1** [comment]
We decided not to backport this one. If this ends up being needed by a customer on RN 62, we will retriage.

**Comment by mjfusa** [comment]
> In the current design we have two levels of Notifications: the top level is in the ReactInstanceSettings.Notifications. They have the same lifespan as the ReactInstanceSettings which is usually given to ReactNativeHost. Then we create a new ReactContext.Notifications that use the ReactInstanceSettings.Notifications as parent. The SendNotification will use them both, while subscribe goes only to ReactContext.Notifications. All ReactContext.Notifications are automatically unsubscribed when we unload the React instance and destroy the ReactContext. This design allows modules and views to talk to each other, and the ReactInstanceSettings.Notifications opens communication with the hosting application.
> 
> Re: Should components from different react contexts be able to communicate?
> 
> * It is not possible in the current implementation. We can add it if we have such scenario.
> 
> In reply to: [631132307](https://github.com/microsoft/react-native-windows/pull/4953#issuecomment-631132307) [](ancestors = 631132307)

Does this mean notifications can't be sent from the app host to a native module? (or vice versa) and only native module to native module notifications are supported at this time?
